src/Workspaces/Core/Portable/Shared/Extensions/IAsyncEnumerableExtensions.cs

```diff
@@ -20,5 +34,51 @@ await foreach (var value in values.WithCancellation(cancellationToken).Configure
 
             return result.ToImmutable();
         }
+
+        /// <summary>
+        /// Takes an array of <see cref="IAsyncEnumerable{T}"/>s and produces a single resultant <see
+        /// cref="IAsyncEnumerable{T}"/> with all their values merged together.  Absolutely no ordering guarantee is
+        /// provided.  It will be expected that the individual values from distinct enumerables will be interleaved
+        /// together.
+        /// </summary>
+        /// <remarks>This helper is useful when doign parallel processing of work where each job returns an <see
+        /// cref="IAsyncEnumerable{T}"/>, but one final stream is desired as the result.</remarks>
+        public static IAsyncEnumerable<T> MergeAsync<T>(this ImmutableArray<IAsyncEnumerable<T>> streams, CancellationToken cancellationToken)
+        {
+            // Code provided by Stephen Toub, but heavily modified after that.
+
+            var channel = Channel.CreateUnbounded<T>();
+
+            var tasks = new Task[streams.Length];
+            for (var i = 0; i < streams.Length; i++)
+                tasks[i] = Process(streams[i], channel.Writer, cancellationToken);
```

Comment:
Is there any chance that one of these streams will be long / infinite and where every MoveNext completed synchronously? If that were to happen, the rest of the method would be blocked behind it, as Process enumerated the whole thing synchronously. It'd also end up OOMing with the channel as an unbounded buffer. You could address that in a variety of ways if it's a real issue, e.g. by changing the channel to be bounded order to apply back pressure.

Reply:
> Is there any chance that one of these streams will be long / infinite

infinite: no.  Long... *possibly*, but extraordinarily unlikely.  But i'll make this bounded just to be on the safe side.

Reply:
done.